manpages/fileserver.1
.TH FILESERVER "7" "October 2012" "CompatibleOne" "Platform"
.SH NAME
fileserver \- CompatibleOne fileserver component.
.SH SYNOPSIS
\fBfileserver\fR
.PP
.SH DESCRIPTION
.\" Add any additional description here
.PP
This will run the CompatibleOne \fBfileserver\fR service.
.SH AUTHOR
The CompatibleOne project.
.SH MANPAGE
Most of this documents comes from the README of ACCORDS, written by Iain James Marshall.
Written by Alexandre Lissy <alissy@mandriva.com> for the CompatibleOne project.
.SH "REPORTING BUGS"
Report CompatibleOne bugs to compatibleone@ow2.org
.br
CompatibleOne home page: <http://www.compatibleone.org>
.SH COPYRIGHT
Copyright \(co 2012 CompatibleOne.
License Apache 2.0.
.br
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.


debian/changelog
compatibleone-accords (0.20121218-1) unstable; urgency=low

  * New upstream code + more doc + sync all binary packages.

 -- Vincent Renardias <vincent@renardias.com>  Tue, 18 Dec 2012 14:43:00 +0200

compatibleone-accords (0.20121214-1) unstable; urgency=low

  * Several bug fixes regarding cosacs.

 -- Vincent Renardias <vincent@renardias.com>  Fri, 14 Dec 2012 16:58:00 +0200

compatibleone-accords (0.20121213-2) unstable; urgency=low

  * Add 'cosacs' package (client side COSACS to be run in the VM images).
  * Split 'compatibleone' in 'compatibleone-base' and 'compatibleone'.

 -- Vincent Renardias <vincent@renardias.com>  Thu, 13 Dec 2012 17:47:00 +0200

compatibleone-accords (0.20121213-1) unstable; urgency=low

  * Major renaming of packages to be fully Debian compliant and prepare the
    integration of the Cosacs client.
  * Force lib # in the dependency and bump the version to make sure we're
    not linking against a version older than today (lib ABI change).

 -- Vincent Renardias <vincent@renardias.com>  Tue, 12 Dec 2012 15:16:00 +0200

compatibleone-accords (0.20121212-1) unstable; urgency=low

  * Upstream update.
  * Major packaging update: each lib now has its own packages.

 -- Vincent Renardias <vincent@renardias.com>  Tue, 11 Dec 2012 15:47:00 +0200

compatibleone-accords (0.20121207-1) unstable; urgency=low

  * Upstream update.
  * Lots of packaging fixes (see gitorious log for the gory details)

 -- Vincent Renardias <vincent@renardias.com>  Fri, 07 Dec 2012 10:25:21 +0200

compatibleone-accords (0.20121129-1) unstable; urgency=low

  * Upstream update.

 -- Vincent Renardias <vincent@renardias.com>  Mon, 26 Nov 2012 11:43:34 +0200

compatibleone-accords (0.20121121-1) unstable; urgency=low

  * Uptream update 2012/11/21 codebase + python enabled.
  * compatibleone-example and compatibleone-testosocci are now gone.

 -- Vincent Renardias <vincent@renardias.com>  Wed, 21 Nov 2012 11:12:59 +0200

compatibleone-accords (0.20121120-1) unstable; urgency=low

  * Uptream update 2012/11/20 codebase + perm fixes.

 -- Vincent Renardias <vrenardias@citypassenger.com>  Tue, 20 Nov 2012 15:24:44 +0200

compatibleone-accords (0.20121114-2) unstable; urgency=low

  * Uptream update 2012/11/19 codebase.

 -- Vincent Renardias <vrenardias@citypassenger.com>  Mon, 19 Nov 2012 20:41:06 +0200

compatibleone-accords (0.20121114-1) unstable; urgency=low

  * Uptream update,
  * jpaasprocci -> paasprocci renaming.
  * fixes for java build on 64bit arch.

 -- Vincent Renardias <vrenardias@citypassenger.com>  Wed, 14 Nov 2012 14:24:57 +0200

compatibleone-accords (0.20121022-1) unstable; urgency=low

  * Upstream change: new accords bash script.
  * All executables now have a manpage.

 -- Vincent Renardias <vrenardias@citypassenger.com>  Mon, 22 Oct 2012 15:18:00 +0200

compatibleone-accords (0.20121015-1) unstable; urgency=low

  * Only recommend compatibleone-paprocci so everybody isn't forced to install the Java runtime.
  * psmisc dep so that we're sure killall is installed.
  * Change OpenSSL dependancy so it also builds on Debian 6.0.x
  * Change debhelper dep so the package builds on Ubuntu 9.04.
  * Commit work with Mauricio to build compatibleone-paprocci and run it with
    the appropriate autodetected java environment.
  * co-stop: can now be run from any directory.
  * Add package compatibleone-example.
  * Add compatibleone-nagios package.

 -- Vincent Renardias <vrenardias@citypassenger.com>  Mon, 15 Oct 2012 11:08:00 +0200

compatibleone-accords (0.20121008-1) unstable; urgency=low

  * More fixes/updates (see gitorious log).

 -- Vincent Renardias <vrenardias@citypassenger.com>  Mon, 8 Oct 2012 14:38:16 +0200

compatibleone-accords (0.20121004-1) unstable; urgency=low

  * More fixes/updates, Initial init script for publisher, WIP.

 -- Vincent Renardias <vrenardias@citypassenger.com>  Thu, 4 Oct 2012 11:55:16 +0200

compatibleone-accords (0.20120918-1) unstable; urgency=low

  * Various fixes/updates, WIP.

 -- Vincent Renardias <vrenardias@citypassenger.com>  Tue, 18 Sep 2012 11:03:43 +0200

compatibleone-accords (0.20120901-1) unstable; urgency=low

  * Initial Debian packaging of CompatibleOne Accords plateform.

 -- Vincent Renardias <vincent@renardias.com>  Sat, 01 Sep 2012 20:33:43 +0200


fileserver/src/fileserver.c
/* ------------------------------------------------------------------- */
/*  ACCORDS PLATFORM                                                   */
/*  (C) 2011 by Iain James Marshall (Prologue) <ijm667@hotmail.com>    */
/* --------------------------------------------------------------------*/
/*  This is free software; you can redistribute it and/or modify it    */
/*  under the terms of the GNU Lesser General Public License as        */
/*  published by the Free Software Foundation; either version 2.1 of   */
/*  the License, or (at your option) any later version.                */
/*                                                                     */
/*  This software is distributed in the hope that it will be useful,   */
/*  but WITHOUT ANY WARRANTY; without even the implied warranty of     */
/*  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU   */
/*  Lesser General Public License for more details.                    */
/*                                                                     */
/*  You should have received a copy of the GNU Lesser General Public   */
/*  License along with this software; if not, write to the Free        */
/*  Software Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA */
/*  02110-1301 USA, or see the FSF site: http://www.fsf.org.           */
/* --------------------------------------------------------------------*/
#ifndef	_fileserver_c
#define	_fileserver_c

#include "standard.h"
#include "fileserver.h"
#include "rest.h"
#include "occi.h"
#include "document.h"
#include "cordspublic.h"

struct	fileserver_configuration Fs = {
	0,
	0,0,0,0,
	(char *) 0,
	(char *) 0,
	_CORDS_DEFAULT_PUBLISHER,
	"admin", "admin",
	"http",  80,
	"domain",
	"fileserver.xml",
	"accords",
	"europe"
	};

public	int	check_debug()		{	return(Fs.debug);		}
public	int	check_verbose()		{	return(Fs.verbose);		}
public	char *	default_publisher()	{	return(Fs.publisher);		}
public	char *	default_tls()		{	return(Fs.tls);			}
public	char *	default_operator()	{	return(Fs.operator);		}
public	char *	default_zone()		{	return(Fs.zone);		}


public	int	failure( int e, char * m1, char * m2 )
{
	if ( e )
	{
		printf("\n*** failure %u",e);
		if ( m1 )
			printf(" : %s",m1);
		if ( m2 )
			printf(" : %s",m2);
		printf(" **`\n");
	}
	return( e );
}

/*	---------------------------------------------	*/  
/*		f i l e s e r v e r _ l o a d 		*/
/*	---------------------------------------------	*/
/*	this function loads fileserver configuration	*/
/*	from the xml configuration file.		*/
/*	---------------------------------------------	*/  
private	void	fileserver_load()
{
	struct	xml_element * document;
	struct	xml_element * eptr;
	struct	xml_element * vptr;
	struct	xml_atribut * aptr;

	set_xml_echo(Fs.debug);

	if (( document = document_parse_file( Fs.config )) != (struct xml_element *) 0)
	{
		if ( Fs.verbose )
			printf("   Broker loading configuration from : %s \n",Fs.config);

		if (( eptr = document_element( document, "fileserver" )) != (struct xml_element *) 0)
		{
			if ((aptr = document_atribut( eptr, "identity" )) != (struct xml_atribut *) 0)
				Fs.identity = document_atribut_string( aptr );
			if ((aptr = document_atribut( eptr, "publisher")) != (struct xml_atribut *) 0)
				Fs.publisher = document_atribut_string( aptr );
			if ((aptr = document_atribut( eptr, "verbose")) != (struct xml_atribut *) 0)
				Fs.verbose = document_atribut_value( aptr );
			if ((aptr = document_atribut( eptr, "debug")) != (struct xml_atribut *) 0)
				Fs.debug = document_atribut_value( aptr );
			if ((aptr = document_atribut( eptr, "threads")) != (struct xml_atribut *) 0)
				Fs.threads = document_atribut_value( aptr );

			if (( vptr = document_element( eptr, "rest" )) != (struct xml_element *) 0)
			{
				if ((aptr = document_atribut( vptr, "host" )) != (struct xml_atribut *) 0)
					Fs.resthost = document_atribut_string( aptr );
				if ((aptr = document_atribut( vptr, "port" )) != (struct xml_atribut *) 0)
					Fs.restport = document_atribut_value( aptr );
			}
			if (( vptr = document_element( eptr, "security" )) != (struct xml_element *) 0)
			{
				if ((aptr = document_atribut( eptr, "authorise")) != (struct xml_atribut *) 0)
					Fs.authorise = document_atribut_value( aptr );
				if ((aptr = document_atribut( vptr, "user" )) != (struct xml_atribut *) 0)
					Fs.user = document_atribut_string( aptr );
				if ((aptr = document_atribut( vptr, "password"))  != (struct xml_atribut *) 0)
					Fs.password = document_atribut_string( aptr );
				if ((aptr = document_atribut( vptr, "tls" )) != (struct xml_atribut *) 0)
					if ((Fs.tls = document_atribut_string( aptr )) != 0)
						if (!( strlen(Fs.tls) ))
							Fs.tls = (char *) 0;
			}
			if (( vptr = document_element( eptr, "domain" )) != (struct xml_element *) 0)
			{
				if ((aptr = document_atribut( vptr, "name" )) != (struct xml_atribut *) 0)
					Fs.domain = document_atribut_string( aptr );
			}
		}
		document = document_drop( document );
	}
	else if ( Fs.verbose )
		printf("   Broker failed to load configuration from : %s \n",Fs.config);
	set_xml_echo(0);
	return;
}

private	int	banner()
{
	printf("\n   CompatibleOne FileServer : Version 1.0a.0.02");
	printf("\n   Beta Version : 14/10/2011");
	printf("\n   Copyright (c) 2011 Iain James Marshall, Prologue");
	printf("\n");
	printf("\nOptions: ");
	printf("\n   --debug                  activates debug messages");
	printf("\n   --verbose                activates info  messages");
	printf("\n   --threads                activate thread control ");
	printf("\n   --config   <file>        load configuration");
	printf("\n   --authorise <value>      set authorization state");            
	printf("\n   --tls      <name>        specify tls configuration");
	printf("\n   --publisher <name>       specify publisher url ");
	printf("\n   --resthost <name>        specify rest server host name or address");
	printf("\n   --restport <number>      specify rest server port");
	printf("\n   --user     <name>        specify user log name   ");
	printf("\n   --password <value>       specify user password   ");
	printf("\n\n");
	return(0);

}

/*	------------------------------------------------------------------	*/
/*			f i l e s e r v e r_ i n i t i a l i s e			*/
/*	------------------------------------------------------------------	*/
private	struct rest_server * fileserver_initialise(  void * v,struct rest_server * sptr )
{
	struct	rest_extension * xptr;
	if (!( xptr = rest_add_extension( sptr ) ))
		return((struct rest_server *) 0);
	else
	{
		xptr->net = (struct connection *) 0;
		return( sptr );
	}
}

/*	------------------------------------------------------------------	*/
/*			f i l e s e r v e r_ a u t h o r i s e 			*/
/*	------------------------------------------------------------------	*/
private	int	fileserver_authorise(  void * v,struct rest_client * cptr, char * username, char * password)
{
	if ( check_verbose )
		printf("   REST Authentication of %s \n",username); 
	if (!( Fs.authorise ))
		return(1);
	else if (!( Fs.user ))
		return(1);
	else if ( strcmp( username, Fs.user ) )
		return(0);
	else if ( strcmp( password, Fs.password ) )
		return(0);
	else if (!( cptr->user = allocate_string( username ) ))
		return(0);
	else if (!( cptr->pass = allocate_string( password ) ))
		return(0);
	else	return(1);
}

/*	------------------------------------------------------------------	*/
/*			f i l e s e r v e r_ e x t e n s i o n 			*/
/*	------------------------------------------------------------------	*/
private	struct rest_extension * fileserver_extension( void * v,struct rest_server * sptr, 
struct rest_extension * xptr)
{
	return( xptr );
}

/*	------------------------------------------------------------------	*/
/*		f i l e s e r v e r_ g e t _ c a p a c i t i e s			*/
/*	------------------------------------------------------------------	*/
private	struct rest_response * fileserver_get_capacities( 
		struct rest_client * cptr, 
		struct rest_request * rptr,
		struct rest_response * aptr )
{
	return( rest_html_response( aptr, 500, "Server Failure" ) );
}

/*	------------------------------------------------------------------	*/
/*			f i l e s e r v e r_ g e t _ f i l e				*/
/*	------------------------------------------------------------------	*/
private	struct rest_response * fileserver_get_file( 
		struct rest_client * cptr, 
		struct rest_request * rptr,
		struct rest_response * aptr )
{
	struct	stat info;
	char	length[64];
	char	buffer[1024];
	char *	nptr;
	if ( stat( rptr->object,&info ) < 0 )
		return( rest_html_response( aptr, 404, "File Not Found" ) );
	else
	{
		sprintf(length,"%u",info.st_size);
		if (!( rest_response_header(aptr,"Content-Length",length) ))
			return( rest_html_response( aptr, 500, "Server Failure" ) );
		else if (!( rest_response_header(aptr,"Content-Type","text/html") ))
			return( rest_html_response( aptr, 500, "Server Failure" ) );
		else if (!( nptr = allocate_string( rptr->object ) ))
			return( rest_html_response( aptr, 500, "Server Failure" ) );
		else
		{
			rest_response_body(aptr,nptr,0 );
			return( rest_response_status( aptr, 200, "OK" ) );
		}
	}
}

/*	------------------------------------------------------------------	*/
/*				f i l e s e r v e r_ g e t 				*/
/*	------------------------------------------------------------------	*/
private	struct rest_response * fileserver_get( void * v,struct rest_client * cptr, struct rest_request * rptr )
{
	struct	rest_response * aptr;
	printf("   Fs GET Request : %s %s %s \n",rptr->method,rptr->object,rptr->version);
	if (!( aptr = rest_allocate_response(cptr) ))
		return( aptr );
	else if ( rptr->body  )
		return( rest_html_response( aptr, 400, "Unexpected Request Body" ) );
	else	return( fileserver_get_file( cptr, rptr, aptr ) );
}

/*	------------------------------------------------------------------	*/
/*				f i l e s e r v e r_ p o s t 				*/
/*	------------------------------------------------------------------	*/
private	struct rest_response * fileserver_post(  void * v,struct rest_client * cptr, 
		struct rest_request * rptr )
{
	struct	rest_response * aptr;
	struct	rest_header   * hptr;
	struct	xml_element   * document;
	FILE 		      * target;
	
	printf("   Fs POST Request : %s %s %s \n",rptr->method,rptr->object,rptr->version);
	if (!( aptr = rest_allocate_response(cptr) ))
		return( aptr );
	else if (!( rptr->body  ))
		return( rest_html_response( aptr, 404, "Expected Body" ) );
	else 	return( rest_html_response( aptr, 400, "Post Not Available" ) );
}

/*	------------------------------------------------------------------	*/
/*				f i l e s e r v e r_ p u t				*/
/*	------------------------------------------------------------------	*/	
private	struct rest_response * fileserver_put(  void * v,struct rest_client * cptr, 
		struct rest_request * rptr )
{
	struct	rest_response * aptr;
	printf("   Fs PUT Request : %s %s %s \n",rptr->method,rptr->object,rptr->version);
	if (!( aptr = rest_allocate_response(cptr) ))
		return( aptr );
	else if (!( rptr->body ))
		return( rest_html_response( aptr, 400, "Put Requires a Body" ) );
	else 	return( rest_html_response( aptr, 400, "Put Not Available" ) );
}

/*	------------------------------------------------------------------	*/
/*				f i l e s e r v e r_ d e l e t e			*/
/*	------------------------------------------------------------------	*/
private	struct rest_response * fileserver_delete(  void * v,struct rest_client * cptr, 
		struct rest_request * rptr )
{
	struct	rest_response * aptr;
	printf("   Fs DELETE Request : %s %s %s \n",rptr->method,rptr->object,rptr->version);
	if (!( aptr = rest_allocate_response(cptr) ))
		return( aptr );
	else if ( rptr->body  )
		return( rest_html_response( aptr, 400, "Unexpected Request Body" ) );
	else 	return( rest_html_response( aptr, 400, "Delete Not Possible" ) );
}

/*	------------------------------------------------------------------	*/
/*				f i l e s e r v e r_ h e a d				*/
/*	------------------------------------------------------------------	*/
private	struct rest_response * fileserver_head(  void * v,struct rest_client * cptr, struct rest_request * rptr )
{
	struct	rest_response * aptr;
	printf("   Fs HEAD Request : %s %s %s \n",rptr->method,rptr->object,rptr->version);
	if (!( aptr = rest_allocate_response(cptr) ))
		return( aptr );
	else if ( rptr->body  )
		return( rest_html_response( aptr, 400, "Unexpected Request Body" ) );
	else 	return( rest_html_response( aptr, 400, "Head Not Possible" ) );
}

/*	------------------------------------------------------------------	*/
/*			f i l e s e r v e r_ o p e r a t i o n				*/
/*	------------------------------------------------------------------	*/
private	int	fileserver_operation( char * nptr )
{
	int	status=0;
	struct	rest_interface  FsI = 
	{
		(void *) 0,
		fileserver_initialise,
		fileserver_authorise,
		fileserver_get,
		fileserver_post,
		fileserver_put,
		fileserver_delete,
		fileserver_head,
		fileserver_extension,
		(void *) 0,
		(void *) 0,
		(void *) 0,
		(void *) 0

	};

	if (!( Fs.authorise ))
		FsI.authorise = (void *) 0;

	return( rest_server(  nptr, Fs.restport, Fs.tls, Fs.threads, & FsI ) );
}


/*	------------------------------------------------------------------	*/
/*				f i l e s e r v e r					*/
/*	------------------------------------------------------------------	*/
private	int	fileserver(int argc, char * argv[] )
{
	int	status=0;
	int	argi=0;
	char *	aptr;
	fileserver_load();
	while ( argi < argc )
	{
		if (!( aptr = argv[++argi] ))
			break;
		else if ( *aptr == '-' )
		{
			aptr++;
			switch( *(aptr++) )
			{
			case	'v'	:
				Fs.verbose=1;
				continue;
			case	'd'	:
				Fs.debug = 0xFFFF;
				continue;
			case	'-'	:
				if (!( strcmp( aptr, "verbose" ) ))
					Fs.verbose = 1;
				else if (!( strcmp( aptr, "debug" ) ))
					Fs.debug = 1;
				else if (!( strcmp( aptr, "tls" ) ))
					Fs.tls = allocate_string( argv[++argi] );
				else if (!( strcmp( aptr, "config" ) ))
				{
					Fs.config = allocate_string( argv[++argi] );
					fileserver_load();
				}
				else if (!( strcmp( aptr, "authorise" ) ))
					Fs.authorise = atoi( argv[++argi] );
				else if (!( strcmp( aptr, "domain" ) ))
					Fs.domain = allocate_string( argv[++argi] );
				else if (!( strcmp( aptr, "user" ) ))
					Fs.user = allocate_string( argv[++argi] );
				else if (!( strcmp( aptr, "password" ) ))
					Fs.password = allocate_string( argv[++argi] );
				else if (!( strcmp( aptr, "resthost" ) ))
					Fs.resthost = allocate_string( argv[++argi] );
				else if (!( strcmp( aptr, "restport" ) ))
					Fs.restport = atoi( argv[++argi] );
				else if (!( strcmp( aptr, "threads" ) ))
					Fs.threads = _REST_THREAD_WORKER;
				else	break;
				continue;
			}
			status = 30;
			break;
		}
		else if (!( status = fileserver_operation(aptr) ))
			continue;
		else	break;
	}
	return(status);
}

/*	------------------------------------------------------------------	*/
/*					m a i n 				*/
/*	------------------------------------------------------------------	*/
public	int	main(int argc, char * argv[] )
{
	if ( argc == 1 )
		return( banner() );
	else	return( fileserver( argc, argv ) );
}


	/* --------- */
#endif	/* _fileserver_c */
	/* --------- */

